pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -278,49 +524,242 @@ func getRequirements(solveNode map[string]interface{}) ([]Requirement, error) {
+			}
+		}
+	}
+
+	for _, arg := range e.getSolveNode().Arguments {
+		if arg.Assignment == nil {
+			continue
+		}
+
+		if arg.Assignment.Name == RequirementsKey {
+			arg.Assignment.Value.List = &requirementsNode
+		}
+	}
+
+	return nil
+}
+
+func (e *BuildExpression) updateRequirement(requirement Requirement) error {
+	if requirement.VersionRequirement == nil {
+		return nil
+	}
+
+	requirementsNode := e.getRequirementsNode()
+
+	for _, r := range requirementsNode {
+		if r.Object == nil {
 			continue
 		}
 
-		if k == funcName {
-			return node, nil
+		for _, o := range *r.Object {
+			if o.Name != RequirementNameKey || *o.Value.Str != requirement.Name {
+				continue
+			}
+
+			var versionRequirements []*Value
+			for _, v := range *r.Object {
+				if v.Name != RequirementVersionRequirementsKey {
+					continue
+				}
+
+				for _, versionReq := range requirement.VersionRequirement {
+					versionRequirements = append(versionRequirements, &Value{Object: &[]*Var{
+						{Name: RequirementComparatorKey, Value: &Value{Str: p.StrP(versionReq[RequirementComparatorKey])}},
+						{Name: RequirementVersionKey, Value: &Value{Str: p.StrP(versionReq[RequirementVersionKey])}},
+					}})
+				}
+				v.Value.List = &versionRequirements
+			}
+
+		}
+	}
+
+	for _, arg := range e.getSolveNode().Arguments {
+		if arg.Assignment == nil {
+			continue
 		}
 
-		// We recurse the build expression until we find the function node
-		if childNode, err := getFuncNode(node, funcName); err == nil {
-			return childNode, nil
+		if arg.Assignment.Name == RequirementsKey {
+			arg.Assignment.Value.List = &requirementsNode
 		}
 	}
 
-	return nil, funcNodeNotFoundError
+	return nil
+}
+
+func (e *BuildExpression) updateTimestamp(timestamp strfmt.DateTime) error {
+	formatted, err := time.Parse(time.RFC3339, timestamp.String())
+	if err != nil {
+		return errs.Wrap(err, "Could not parse latest timestamp")
+	}
+
+	for _, arg := range e.getSolveNode().Arguments {
+		if arg.Assignment == nil {
+			continue
+		}
+
+		if arg.Assignment.Name == "at_time" {
+			arg.Assignment.Value.Str = p.StrP(formatted.Format(time.RFC3339))
+		}
+	}
+
+	return nil
+}
+
+func (e *BuildExpression) MarshalJSON() ([]byte, error) {
+	m := make(map[string]interface{})
+	let := make(map[string]interface{})
+	for _, assignment := range e.Let.Assignments {
+		let[assignment.Name] = assignment.Value
+	}
+	m["let"] = let
+	return json.Marshal(m)
+}
+
+func (a *Var) MarshalJSON() ([]byte, error) {
+	m := make(map[string]interface{})
+	m[a.Name] = a.Value
+	return json.Marshal(m)
+}
+
+func (v *Value) MarshalJSON() ([]byte, error) {
+	switch {
+	case v.Ap != nil:
+		return json.Marshal(v.Ap)
+	case v.List != nil:
+		return json.Marshal(v.List)
+	case v.Str != nil:
+		return json.Marshal(strings.Trim(*v.Str, `"`))
+	case v.Null != nil:
+		return json.Marshal(nil)
+	case v.Assignment != nil:
+		return json.Marshal(v.Assignment)
```

Comment:
Can't these be addressed with an `OmitEmpty`?

Reply:
No, we're not unmarshaling to these types. We first unmarshal to a map and then the "parser" iterates through the map to populate these values.